rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+slippery concept without concrete examples (at least that's what I found when
+writing this document).
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+origin_index_result_ids - filtered_index_result_ids
+# => [1, 2, 3, 4] - [1] = [2, 3, 4]
+```
+
+Note how the final resulting list of IDs includes results marked mature, but
+that do not have sensitive textual content (result 2). Unfortunately there is no
+way around this as if we were to exclude the results marked `mature` from the
+origin result IDs in the final set subtraction, we'd end up with the following
+situation:
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+origin_index_less_mature_result_ids = [r for r in origin_index results if not r["mature"]]
+# => [1, 3]
+
+origin_index_less_mature_result_ids - filtered_index_result_ids
+# => [1, 3] - [1] = [3]
+```
+
+Notice how we still have a problem: we've excluded the result that has sensitive
+text _and_ is marked mature (result 4).
+
+It's tempting to think we could move the mature filtering to _after_ when we've
+derived the list of sensitive result IDs. However, the problem will remain the
+same: we will exclude results that are both marked mature and have sensitive
+textual content:
+
+```
+origin_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+  {id: 2, mature: true}, # no sensitive text
+  {id: 3, mature: false}, # has sensitive text
+  {id: 4, mature: true}, # has sensitive text
+]
+
+filtered_index_results = [
+  {id: 1, mature: false}, # no sensitive text
+]
+
+sensitive_result_ids = origin_index_result_ids - filtered_index_result_ids
+# => [1, 2, 3, 4] - [1] = [2, 3, 4]
+
+sensitive_results = [r for r in origin_index_results if r["_id"] in sensitive_result_ids and not r["mature"]]
+# => [3]
+```
+
+Result 4 is still excluded because we can't tell in the final operation whether
+result 4 was not in the set of IDs from the filtered index _only_ because it is
+marked mature or if it also has sensitive textual content.
+
+The only way I can think of to solve this is to have an index solely of
+documents that _do_ have sensitive textual content and querying it to derive the
+final list of sensitive results without needing to rely on the diversely
+filtered index. However, this adds a big re-indexing burden (yet another
+additional index) that can be bypassed altogether if we just exclude mature
+results from the filtered index to begin with.
```

Comment:
Forgive me if I'm missing a piece here - could we exclude mature results from the filtered index as well when we're creating it? Would there be a need to include mature results that aren't sensitive separately, especially since we're going to be excluding those results anyway if the sensitive/mature results aren't requested to be included in the final results?

Reply:
> Forgive me if I'm missing a piece here - could we exclude mature results from the filtered index as well when we're creating it? Would there be a need to include mature results that aren't sensitive separately, especially since we're going to be excluding those results anyway if the sensitive/mature results aren't requested to be included in the final results?

As Staci said, the section deals entirely with the idea of excluding mature documents from the filtered index and describes why it's an impossible solution. @AetherUnbound, your inclination is precisely why I wrote the section this paragraph is under, because I also considered it a possibility and anticipated that others would have the same question. If the explanation is not clear on why it's not possible for us to exclude the mature results from the filtered index, please let me know so that I can revise the section.

Reply:
Admittedly I read this section several times and still came away a bit confused 😅 I think the premise was getting to me - my thought was that it doesn't matter for the API whether the `mature` field is `True` or whether we filtered it for sensitive text reasons, the results are excluded all the same. So _in my mind_ making the distinction between `mature` and `sensitive` was unnecessary, but it seems that it might be in some cases which would require the explanation you put out Sara. I guess the core of my question was whether that distinction was important, although you've [answered my question in a later comment](https://github.com/WordPress/openverse/pull/996#discussion_r1162130216)!

I perhaps was getting confused by the distinction between the `mature` API parameter (which we are saying is functionally equivalent to `sensitive` after the implementation of this plan) and the `mature` _field_ on the API (which will stick around in perpetuity) 🤯 I believe I'm all cleared up now!

Reply:
> I perhaps was getting confused by the distinction between the mature API parameter (which we are saying is functionally equivalent to sensitive after the implementation of this plan) and the mature field on the API (which will stick around in perpetuity) exploding_head I believe I'm all cleared up now!

This will be confusing for a while, unfortunately :disappointed:. Hopefully decoupling the parameter name from the various things that it includes will help (i.e., using the broader "sensitive" term to capture "mature" results as well as other types of sensitivity, textual, and machine vision derived).

> So in my mind making the distinction between mature and sensitive was unnecessary, but it seems that it might be in some cases which would require the explanation you put out Sara

I can add another note about this in the section referencing the frontend. The existence of the `sensitivity` field as an array that notes the distinction hopefully communicates the necessity of knowing the difference between them. The reason for that need isn't explicit in this document because the project proposal covers the reason we are making this distinction.